relay/pkg/block/forwarder.go

```diff
@@ -0,0 +1,110 @@
+package block
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ipfs/go-log"
+	"github.com/keep-network/tbtc/relay/pkg/btc"
+	"github.com/keep-network/tbtc/relay/pkg/chain"
+)
+
+const (
+	// Size of the headers queue.
+	headersQueueSize = 50
+
+	// Maximum size of processed headers batch.
+	headersBatchSize = 5
+
+	// Maximum time for which the pulling process will wait for a single header
+	// to be delivered by the headers queue.
+	headerTimeout = 1 * time.Second
+
+	// Block duration of a Bitcoin difficulty epoch.
+	difficultyEpochDuration = 2016
+
+	// Duration for which the forwarder should rest after performing
+	// a push action.
+	forwarderPushingSleepTime = 45 * time.Second
+)
+
+var logger = log.Logger("relay-block-forwarder")
+
+// Forwarder takes blocks from the Bitcoin chain and forwards them to the
+// given host chain.
+type Forwarder struct {
+	btcChain  btc.Handle
+	hostChain chain.Handle
+
+	processedHeaders int
+
+	headersQueue chan *btc.Header
+	errChan      chan error
+}
+
+// RunForwarder creates an instance of the block forwarder and runs its
+// processing loop. The lifecycle of the forwarder loop can be managed
+// using the passed context.
+func RunForwarder(
+	ctx context.Context,
+	btcChain btc.Handle,
+	hostChain chain.Handle,
+) *Forwarder {
+	forwarder := &Forwarder{
+		btcChain:     btcChain,
+		hostChain:    hostChain,
+		headersQueue: make(chan *btc.Header, headersQueueSize),
+		errChan:      make(chan error, 1),
+	}
+
+	go forwarder.pushingLoop(ctx)
+
+	return forwarder
+}
+
+func (f *Forwarder) pushingLoop(ctx context.Context) {
+	logger.Infof("running forwarder pushing loop")
+
+	for {
+		select {
+		case <-ctx.Done():
+			logger.Infof("forwarder context is done")
```

Comment:
I don't think we need it, especially that it is already logged one level higher. We can log it here or in the code calling `RunForwarder` but let's do it only one time.

Reply:
Done in https://github.com/keep-network/tbtc/commit/c6fb5d71e9c976fb98ee676e36dfcd48ff6221fe